packages/contractkit/src/wallets/signers/azure-key-vault-client.ts

```diff
@@ -0,0 +1,191 @@
+    if (!AzureKeyVaultClient.isCanonical(S, N)) {
+      S = secp256k1Curve.curve.n.sub(S)
+    }
+
+    const rBuff = AzureKeyVaultClient.bigNumberToBuffer(R)
+    const sBuff = AzureKeyVaultClient.bigNumberToBuffer(S)
+    const canonicalizedSignature = Buffer.concat([rBuff, sBuff])
+    const publicKey = await this.getPublicKey(keyName)
+
+    // Azure doesn't provide the recovery key in the signature
+    const recoveryParam = AzureKeyVaultClient.recoverKeyIndex(
+      canonicalizedSignature,
+      publicKey,
+      message
+    )
+    return new Signature(recoveryParam, rBuff, sBuff)
+  }
+
+  public async hasKey(keyName: string): Promise<boolean> {
+    try {
+      await this.keyClient.getKey(keyName)
+    } catch (e) {
+      if (e.message.includes('this is not a valid private key')) {
+        return false
+      }
+      throw e
+    }
+    return true
+  }
+
+  /**
+   * Returns true if the signature is in the "bottom" of the curve
+   */
+  private static isCanonical(S: BigNumber, curveN: BigNumber): boolean {
+    return S.comparedTo(curveN.shiftedBy(1)) <= 0
+  }
+
+  /**
+   * Attempts each recovery key to find a match
+   */
+  private static recoverKeyIndex(
+    signature: Uint8Array,
+    publicKey: BigNumber,
+    hash: Uint8Array
+  ): number {
+    for (let i = 0; i < 4; i++) {
+      const compressed = false
+      try {
+        const recoveredPublicKeyByteArr = ecdsaRecover(signature, i, hash, compressed)
+        const publicKeyBuff = Buffer.from(recoveredPublicKeyByteArr)
+        const recoveredPublicKey = AzureKeyVaultClient.bufferToBigNumber(publicKeyBuff)
+        if (publicKey.eq(recoveredPublicKey)) {
+          return i
+        }
+      } catch (e) {
+        if (e.message === 'Public key could not be recover') {
+          throw new Error('Invalid public key. Ensure that the key type is ECDSA')
+        }
+      }
+    }
+    throw new Error('Unable to generate recovery key from signature.')
+  }
+
+  private async getKey(keyName: string): Promise<KeyVaultKey> {
+    try {
+      const signingKey = await this.keyClient.getKey(keyName)
+
+      if (
+        typeof signingKey?.id === 'undefined' ||
+        typeof signingKey?.key?.x === 'undefined' ||
+        typeof signingKey?.key?.y === 'undefined'
+      ) {
+        throw new Error(`Invalid key data returned from Azure: ${signingKey}`)
+      }
+
+      return signingKey
+    } catch (e) {
+      if (e.message.includes('Key not found')) {
+        throw new Error(`Key ${keyName} not found in KeyVault ${this.vaultName}`)
+      }
+      throw new Error(`Unexpected KeyVault error ${e.message}`)
+    }
+  }
+
+  private static bufferToBigNumber(input: Buffer): BigNumber {
+    return new BigNumber(ensureLeading0x(input.toString('hex')))
+  }
+  private static bigNumberToBuffer(input: BigNumber): Buffer {
+    return ethUtil.toBuffer(ensureLeading0x(input.toString(16))) as Buffer
+  }
+}
+
+export class Signature {
```

Comment:
Seems like something that's useful outside of `azure-key-vault-client.ts`, maybe it should live somewhere else?

Reply:
Agree. I think this is part of larger work to consolidate the signing libraries/utilities. There exist some similar objects throughout the repo, but they don't provide much guarantees around the contents. I've added an issue #3410 to track this work.